pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+// It is also called when the IP is upserted with a new identity
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, ip net.IP) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIp[identity.ID]; ok {
+		newIps := slices.DeleteFunc(ips, func(existing net.IP) bool {
+			return existing.Equal(ip)
+		})
+		if len(newIps) == 0 {
+			delete(s.currentIdentityToIp, identity.ID)
+		} else {
+			s.currentIdentityToIp[identity.ID] = newIps
+		}
+	}
+	return nil
+}
+
+// UpdatePolicyRulesLocked updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+// The UpdatePolicyRulesLocked method is called with the proxy.mutex lock held
+func (s *FQDNDataServer) UpdatePolicyRulesLocked(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	s.snapshotMutex.Lock()
+	defer s.snapshotMutex.Unlock()
+
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
+		Response: pb.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// Starts the Standalone DNS Proxy grpc server on the given port
+func (s *FQDNDataServer) Start() error {
+	address := fmt.Sprintf("localhost:%d", s.port)
+	s.log.Info("Starting Standalone DNS Proxy server on: ", logfields.Address, address)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		s.log.Error("Failed to listen: ", logfields.Error, err)
+		return err
+	}
+	grpcServer := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(kaep), grpc.KeepaliveParams(kasp))
+	s.grpcServer = grpcServer
+
+	pb.RegisterFQDNDataServer(grpcServer, s)
+
+	if err := s.grpcServer.Serve(lis); err != nil {
+		s.log.Error("Failed to serve: ", logfields.Error, err)
+		return err
+	}
+
+	return nil
+}
+
+func (s *FQDNDataServer) Stop() {
+	if s.grpcServer == nil {
+		s.log.Error("GRPC server is nil, cannot stop")
+		return
+	}
+	s.log.Info("Stopping Standalone DNS Proxy server")
```

Comment:
This sort of `Info` message isn't necessary. Hive already logs when it calls stop hooks.